### PR TITLE
SLING-10110: added a functionality for resolving Star-Wildcards in ig…

### PR DESCRIPTION
**Pull request by anban02**
Hello everyone

about a year ago I opened the Issue [SLING-10110](https://issues.apache.org/jira/projects/SLING/issues/SLING-10110?filter=allopenissues). Now after a year i thought that I'd try to add the functionality for resolving Star-Wildcards myself.

The idea here is that you can copy the ignoreUrlParams as they are from the Apache-Dispatcher-Config and add them to your slingDynamicInclude-config.

There might be some Indent-Changes etc. that are probably not conform to the project. If there are any Code-Style Guidelines please tell me and I'll restyle the code as needed.

Please also let me know if there is anything in my added code that does not perform well or could be done better. It was the best solution i could find at the moment but if someone has any ideas for improvements please let me know and I'll try to implement them.

I also used Java8 Streaming-API but I'm not quite sure that the project runs above Java7. Maybe it only looked like that because some code-snippets looked a bit older to me. Please don't misunderstand this as something negative. It is only my personal perception.

Important: If this change is accepted the documentation of ignoreUrlParams would need to be updated.
In the readMe it points to this URL that is apparently dead: https://docs.adobe.com/docs/en/dispatcher/disp-config.html#Ignoring%20URL%20Parameters 

Thank you in advance for reviewing this.

### CODE REVIEW

**Comment by anban02** [comment]
I just made some adjustments to my code and trimmed it down so it is easier to read and doesn't do this hard to understand Collection-Modification.

I think it should be much easier to read and understand than it was with my initial commit.

**Comment by joerghoh** [comment]
@anban02 So when looking at the patch, it just translates a glob ```*``` into its equivalent regex counterpart ```.*```; is this the gist of the patch?

**Comment by anban02** [comment]
> @anban02 So when looking at the patch, it just translates a glob `*` into its equivalent regex counterpart `.*`; is this the gist of the patch?

That is not all but a big part of it.
In the current unpatched version the ignoreUrlParams matching is an exact value matching that did not allow the use of wildcards.
But because Apache-Dispatcher Configurations allow wildcards I needed SDI to also correctly resolve ignoreUrlParams that were defined with wildcards.

So to summarize it:
i added a check for ignoreUrlParams to see if there is a wildcard configured in the style of a dispatcher config. If so then the `*` is translated to `.*` so i can do a regex match for that and see if the parameter that is currently checked should be ignored.

**Comment by rombert** [comment]
@anban02 - I am unfortunately quite lost - SDI and the Dispatcher are not at all known to me. @joerghoh - what are your thoughts on this PR?

**Comment by joerghoh** [comment]
@anban02 So if I understand this correctly, you want to extend SDI to ignore url parameters not only by exact match (what is already supported today), but you want to add support for globbed URL parameters (for example "utm*"). 

If you internally translate such a glob already to a regex, why not support directly regexes?

**Comment by anban02** [comment]
> @anban02 So if I understand this correctly, you want to extend SDI to ignore url parameters not only by exact match (what is already supported today), but you want to add support for globbed URL parameters (for example "utm*").
> 
> If you internally translate such a glob already to a regex, why not support directly regexes?

Yes. that is correct. In the use case i have at my company there are some of these globbed parameters for tracking, as tracking-parameters can be anything they want and just prefix with a certain kind of String like for example "tra-*".

How would one go about directly supporting regex here? Can you give me an example for that one as i don't seem to understand what exactly that implies in this scenario.

File: src/main/java/org/apache/sling/dynamicinclude/util/RequestHelperUtil.java [hunk 1]
```diff
@@ -0,0 +1,55 @@
...
+
+	private RequestHelperUtil() {
+		// private constructor to prevent instance creation of util classes
+	}
+
+	/**
+	 * Checks if a request contains any parameters that are not defined in the ignoreUrlParams.
+	 * Wildcards as they are possible to configure on Apache-Dispatcher, are also properly checked.
+	 *
+	 * @param ignoreUrlParams The list of configured ignoreUrlParams (ignoreUrlParams can be defined with regex-pattern)
```
Comment (joerghoh) [inline src/main/java/org/apache/sling/dynamicinclude/util/RequestHelperUtil.java:36]:
```suggestion
	 * @param ignoreUrlParams The list of configured ignoreUrlParams (regular expressions)
```

(they *can* not only be regex-patterns, they always are!)

File: src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java [hunk 2]
```diff
@@ -0,0 +1,143 @@
...
+	private SlingHttpServletRequest slingHttpServletRequest;
+
+	@Before
+	public void prepareTestData() {
+		slingHttpServletRequest = Mockito.mock(SlingHttpServletRequest.class);
+	}
+
+	@Test
+	public void requestHasParameters_noParametersAtAll() {
+		Collection<String> ignoreUrlParams = new ArrayList<>();
```
Comment (joerghoh) [inline src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java:50]: Every test method has this statement, so you could turn it into a class member and re-initialize it in a @Before method.

File: src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java [hunk 3]
```diff
@@ -0,0 +1,143 @@
...
+
+		Mockito.when(slingHttpServletRequest.getParameterMap()).thenReturn(new HashMap<>());
```
Comment (joerghoh) [inline src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java:52]: same with this statement. Create another global property, and reset it in the @Before method, and then you can just use it in every Test method.

**Review by joerghoh — APPROVED** [review]
lgtm, just minor improvements.

**Comment by anban02** [comment]
I just updated the code but now that it is less code and because of the change in the IncludeTagFilter to make it call the Util-Class I don't have the mandatory 80% test coverage, because the IncludeTagFilter never had any Unit-Tests to begin with.

I'll try to write some useful tests but i have never written Unit-Tests for javax.servlet.Filter implementations.

Is Bertrand Delacretaz still active in this project? He seems to have done most of the implementation in that class so maybe he could give me some pointers or any other kind of help.

File: src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java (hunk 3 above)
Comment (anban02) [reply src/test/java/org/apache/sling/dynamicinclude/util/RequestHelperUtilTest.java:52]: Thank you. I'll improve on the points you have mentioned and then do a new push.

**Review by joerghoh — APPROVED** [review]

**Comment by joerghoh** [comment]
thanks @anban02 for your contribution, I merged it now. Let's discuss the release over in SLING-10110.

**Pull request merged**